File: javascript/src/credits/service-types.ts
import type Decimal from "decimal.js";

import type { UsageMetrics } from "../metrics.js";
import type { Logger } from "../shared/logger.js";
import type { CreditEvent } from "./events.js";
import type {
  BillingMode,
  CreditMetadata,
  DeductionResult,
  DeductionSuccess,
  UsageAnalyticsStore,
  UsageChargeStore,
} from "./types/index.js";

export type PolicyPreset = "strict_prepaid" | "overdraft";
export type MetricsOrAmount = UsageMetrics | Decimal | number;

export interface PostDeductionContext {
  userId: string;
  source: "deduct" | "settle" | "raw";
  deduction: DeductionSuccess;
}

export interface LowBalanceConfig {
  thresholds?: (Decimal | number)[] | null;
  onTrigger?: ((event: CreditEvent) => void | Promise<void>) | null;
  /**
   * Maximum number of subjects whose breached-threshold state is retained in
   * memory. Defaults to 100,000.
   */
  maxTrackedUsers?: number;
}

export interface CreditsServiceOptions {
  logger?: Logger | null;
  /**
   * Optional read-only analytics backend. Defaults to the credit store, which
   * keeps PostgreSQL as the zero-infrastructure behavior.
   */
  analytics?: UsageAnalyticsStore | null;
  /** Optional usage-history backend. Defaults to the credit store. */
  usageStore?: UsageChargeStore | null;
  /** Fallback credit policy for subjects without a plan assignment. */
  policy?: PolicyPreset;
  /** Planless-subject floor used with the `overdraft` fallback. */
  overdraftFloor?: Decimal | number | null;
  /** Planless-subject admission limit fallback. */
  maxConcurrent?: number | null;
  /**
   * Edge-triggered low-balance thresholds and optional non-blocking handler.
   * Defaults to zero when omitted.
   */
  lowBalance?: LowBalanceConfig | null;
  /** Default lease TTL in seconds. Defaults to 600. */
  defaultTtlSeconds?: number;
  /** Sweep a subject's expired credits before balance-sensitive operations. */
  lazyExpiry?: boolean;
  /**
   * Catalog cache TTL in milliseconds. A value of 0 disables automatic
   * reloads. Concurrent reloads are deduplicated. Defaults to 300,000.
   */
  catalogCacheTtlMs?: number;
  /**
   * Awaited after a committed, non-replayed deduction. Hook failures are
   * isolated from the committed credit charge.
   */
  postDeduction?: ((context: PostDeductionContext) => void | Promise<void>) | null;
}

export interface AddCreditsOptions {
  type?: string;
  metadata?: CreditMetadata | null;
  expiresAt?: Date | null;
  /** Target credit bucket; omitted resolves to the catalog's default bucket. */
  bucket?: string | null;
  /** Stable replay key for the ledger mutation. */
  idempotencyKey?: string | null;
}

export interface DeductCreditsOptions {
  entryType?: string;
  bucket?: string | null;
  metadata?: CreditMetadata | null;
  /** Stable replay key for the ledger mutation. */
  idempotencyKey?: string | null;
}

export interface DeductOptions {
  idempotencyKey?: string | null;
  metadata?: CreditMetadata | null;
  /** Entitlement feature required for this operation. */
  feature?: string | null;
}

export type DeductFlatJobOptions = DeductOptions;

export interface RecordUsageOptions {
  idempotencyKey?: string | null;
  metadata?: CreditMetadata | null;
}

export interface RefundCreditsOptions {
  amount?: Decimal | number;
  reason?: string;
  metadata?: CreditMetadata | null;
  idempotencyKey?: string | null;
}

export interface DeductTeamOptions {
  idempotencyKey?: string | null;
  metadata?: CreditMetadata | null;
}

export interface ReserveOptions {
  /** Replay-safe acquisition key. A random key is generated when omitted. */
  idempotencyKey?: string | null;
  operationType?: string;
  billingMode?: BillingMode | null;
  ttl?: number | null;
  metadata?: CreditMetadata | null;
  feature?: string | null;
  /** Model tag used when reserving a raw amount instead of usage metrics. */
  model?: string | null;
}

export interface SettleOptions {
  idempotencyKey?: string | null;
  metadata?: CreditMetadata | null;
  /** Entitlement key supplied at reserve time. */
  feature?: string | null;
}

export interface CanAffordOptions {
  feature?: string | null;
  billingMode?: BillingMode | null;
  operationType?: string;
}

export interface GrantSubscriptionCycleOptions {
  bucket?: string;
  expiresAt?: Date;
  ttlDays?: number;
  planKey?: string;
  idempotencyKey?: string;
  metadata?: CreditMetadata | null;
}

export interface RunBilledOptions<T> {
  estimate: MetricsOrAmount;
  doWork: () => Promise<{ result: T; actual: MetricsOrAmount }>;
  operationType?: string;
  billingMode?: BillingMode | null;
  /** Stable key for the complete reserve/work/settle operation. */
  operationKey?: string | null;
  ttl?: number | null;
  feature?: string | null;
  metadata?: CreditMetadata | null;
  /** Settlement attempts for SDK-classified transient failures. Defaults to 3. */
  settlementAttempts?: number;
}

export interface BeginBilledOperationOptions {
  estimate: MetricsOrAmount;
  /** Stable key for the complete reserve/settle operation. */
  operationKey: string;
  operationType?: string;
  billingMode?: BillingMode | null;
  ttl?: number | null;
  feature?: string | null;
  metadata?: CreditMetadata | null;
}

export interface BilledOperation {
  readonly leaseId: string;
  readonly operationKey: string;
  settle(actual: MetricsOrAmount, metadata?: CreditMetadata | null): Promise<DeductionResult>;
  renew(ttl?: number | null): Promise<void>;
  release(): Promise<void>;
}

File: python/src/bursar/credits/service_types.py
"""Typed options for :mod:`bursar.credits.service`.

These Pydantic schemas mirror ``javascript/src/credits/service-types.ts`` while
using Python naming conventions.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, SkipValidation, field_validator

from bursar.credits.events import CreditEvent
from bursar.credits.types import (
    BillingMode,
    CreditMetadata,
    DeductionResult,
    UsageAnalyticsStore,
    UsageChargeStore,
)
from bursar.metrics import UsageMetrics
from bursar.shared.idempotency import StableKey
from bursar.shared.logger import Logger

PolicyPreset = Literal["strict_prepaid", "overdraft"]
PostDeductionSource = Literal["deduct", "settle", "raw"]
ExactAmount = Decimal | int | str
MetricsOrAmount = UsageMetrics | ExactAmount


ReplayKey = StableKey


def _parse_exact_amount(value: object, field_name: str) -> Decimal:
    if isinstance(value, Decimal):
        return value
    if isinstance(value, bool) or not isinstance(value, (int, str)):
        raise ValueError(f"{field_name} must be a Decimal, integer, or decimal string")
    try:
        return Decimal(value)
    except ArithmeticError as error:
        raise ValueError(f"{field_name} must be a valid decimal string") from error


class _CreditsServiceModel(BaseModel):
    model_config = ConfigDict(extra="forbid", arbitrary_types_allowed=True)


class PostDeductionContext(_CreditsServiceModel):
    user_id: str
    source: PostDeductionSource
    deduction: DeductionResult


class LowBalanceConfig(_CreditsServiceModel):
    thresholds: list[ExactAmount] | None = None
    on_trigger: Callable[[CreditEvent], None | Awaitable[None]] | None = None
    max_tracked_users: int = Field(default=100_000, strict=True, ge=1)

    @field_validator("thresholds", mode="before")
    @classmethod
    def validate_thresholds(cls, values: object) -> object:
        if values is None or not isinstance(values, (list, tuple)):
            return values
        normalized = [_parse_exact_amount(value, "low-balance threshold") for value in values]
        if any(not value.is_finite() or value < 0 for value in normalized):
            raise ValueError("low-balance thresholds must be finite non-negative amounts")
        return normalized


class CreditsServiceOptions(_CreditsServiceModel):
    logger: SkipValidation[Logger] | None = None
    analytics: UsageAnalyticsStore | None = None
    usage_store: UsageChargeStore | None = None
    policy: PolicyPreset = "strict_prepaid"
    overdraft_floor: ExactAmount | None = None
    max_concurrent: int | None = Field(default=None, strict=True, ge=1)
    low_balance: LowBalanceConfig | None = None
    default_ttl_seconds: int = Field(default=600, strict=True, ge=1)
    lazy_expiry: bool = Field(default=False, strict=True)
    catalog_cache_ttl_ms: int = Field(default=300_000, strict=True, ge=0)
    post_deduction: Callable[[PostDeductionContext], None | Awaitable[None]] | None = None

    @field_validator("overdraft_floor", mode="before")
    @classmethod
    def validate_overdraft_floor(cls, value: object) -> object:
        if value is None:
            return None
        normalized = _parse_exact_amount(value, "overdraft_floor")
        if not normalized.is_finite() or normalized > 0:
            raise ValueError("overdraft_floor must be finite and <= 0")
        return normalized


class ReserveOptions(_CreditsServiceModel):
    idempotency_key: ReplayKey
    operation_type: str | None = None
    billing_mode: BillingMode | None = None
    ttl: int | None = Field(default=None, strict=True, ge=1)
    metadata: CreditMetadata | None = None
    feature: str | None = None
    model: str | None = None


class SettleOptions(_CreditsServiceModel):
    idempotency_key: ReplayKey | None = None
    metadata: CreditMetadata | None = None
    feature: str | None = None


class CanAffordOptions(_CreditsServiceModel):
    feature: str | None = None
    billing_mode: BillingMode | None = None
    operation_type: str = "usage"


class GrantSubscriptionCycleOptions(_CreditsServiceModel):
    bucket: str = "subscription"
    expires_at: datetime | None = None
    ttl_days: int | None = Field(default=None, strict=True, ge=1)
    plan_key: str | None = None
    idempotency_key: ReplayKey
    metadata: CreditMetadata | None = None


class RunBilledOptions(_CreditsServiceModel):
    estimate: MetricsOrAmount
    do_work: Callable[[], tuple[Any, MetricsOrAmount]]
    operation_type: str = "usage"
    billing_mode: BillingMode | None = None
    operation_key: ReplayKey
    ttl: int | None = Field(default=None, strict=True, ge=1)
    feature: str | None = None
    metadata: CreditMetadata | None = None
    settlement_attempts: int = Field(default=3, strict=True, ge=1)


class RunBilledAsyncOptions(_CreditsServiceModel):
    estimate: MetricsOrAmount
    do_work: Callable[[], Awaitable[tuple[Any, MetricsOrAmount]]]
    operation_type: str = "usage"
    billing_mode: BillingMode | None = None
    operation_key: ReplayKey
    ttl: int | None = Field(default=None, strict=True, ge=1)
    feature: str | None = None
    metadata: CreditMetadata | None = None
    settlement_attempts: int = Field(default=3, strict=True, ge=1)


class BeginBilledOperationOptions(_CreditsServiceModel):
    estimate: MetricsOrAmount
    operation_key: ReplayKey
    operation_type: str = "usage"
    billing_mode: BillingMode | None = None
    ttl: int | None = Field(default=None, strict=True, ge=1)
    feature: str | None = None
    metadata: CreditMetadata | None = None


__all__ = [
    "CanAffordOptions",
    "BeginBilledOperationOptions",
    "CreditsServiceOptions",
    "ExactAmount",
    "GrantSubscriptionCycleOptions",
    "LowBalanceConfig",
    "MetricsOrAmount",
    "PolicyPreset",
    "PostDeductionContext",
    "PostDeductionSource",
    "ReserveOptions",
    "RunBilledOptions",
    "RunBilledAsyncOptions",
    "SettleOptions",
]
